src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -295,24 +292,26 @@ private void DisconnectFromWorkspaceIfOpen()
 
                 if (_workspace.IsDocumentOpen(_documentIdentity.DocumentId))
                 {
-                    _workspace.OnSourceGeneratedDocumentClosed(_documentIdentity.DocumentId);
+                    var sourceGeneratedDocument = (SourceGeneratedDocument?)_textBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                    Contract.ThrowIfNull(sourceGeneratedDocument);
+                    _workspace.OnSourceGeneratedDocumentClosed(sourceGeneratedDocument);
                 }
             }
 
             public void Dispose()
             {
                 AssertIsForeground();
 
+                _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+
+                DisconnectFromWorkspaceIfOpen();
```

Comment:
Was there a reason to reorder this earlier? The ordering makes sense, but if there's a functional reason it may be worth commenting before somebody breaks this again.

Reply:
➡️ Now commented